Increment discarded spans metric for all limit reasons

joe-elliott opened the pull request:
**What this PR does**:
`tempo_discarded_spans_total` is meant to be a metric to inform tenants as to why they may be getting refused spans.  Previously it only covered the most basic rate limiting case.   This PR looks to extend that to include all limit based reasons that a span might be refused.

There is unfortunately no 100% correct way to metric this.  For instance, in the provided solution we are only including spans from the one trace that was rejected however the distributor will refuse the entire batch.  

We could try to metric this in the distributor which would make the metric more authentic.  It would be gross to try to get the rejection reason.  Substring matching?  It is also true that if a batch is rejected in the ingester it has already probably succeeded on other traces.

No great answers here.  Suggestions welcome.

joe-elliott commented:
Changed course and decided that counting all discarded spans inside the distributor.  This does require some string matching, but hoping that testing will keep things in sync.

Overall prefer this method as it keeps the discarded spans metricing in one place, allows us to metric other failures inside of tempo, and over counts dropped spans instead of undercounts (which I prefer).

mdisibio commented on modules/distributor/distributor.go:
```diff
@@ -32,9 +34,14 @@ import (
+	// reasonTraceTooLarge indicates that a single trace has too many spans
+	reasonTraceTooLarge = "trace_too_large"
+	// reasonTooManyTraces indicates that tempo is already tracking too many live traces in the ingesters for this user
+	reasonTooManyTraces = "too_many_traces"
```
`too_many_traces` could also be interpreted as a rate limit.  Thoughts on an alternate string like `live_traces_exceeded` ?

mdisibio commented on modules/overrides/limits.go:
```diff
@@ -6,11 +6,15 @@ import (
 	GlobalIngestionRateStrategy = "global"
+
+	// ErrorPrefixTooManyTraces is used to flag batches from the ingester that were rejected b/c they had too many traces
+	ErrorPrefixTooManyTraces = "TOO_MANY_TRACES:"
```
`too_many_traces` could also be interpreted as a rate limit. Thoughts on tweaking it or an alternate string like `live_traces_exceeded` ?

joe-elliott replied:
agree :+1:

joe-elliott commented:
- too_many_traces => live_traces_exceeded
- added test for spans/second rate limit
- removed "internal_error" reason.  Since we can't 100% accurately account for this I'd rather this metric only represent those spans that were discarded due to limits.

annanay25 commented on modules/distributor/distributor.go:
```diff
@@ -347,3 +355,17 @@ func requestsByTraceID(req *tempopb.PushRequest, userID string, spanCount int) (
+	}
+	desc := s.Message()
+
+	if strings.HasPrefix(desc, overrides.ErrorPrefixLiveTracesExceeded) {
```
nit: Metric'ing based on an error message prefix from a different service seems fragile.. We could wrap the error or change it in many different ways in the future. Should we change this to a `strings.Contains()`? Or should we map different codes in the `codes` package to our errors?

joe-elliott replied:
> Metric'ing based on an error message prefix from a different service seems fragile..
We have tests to detect drift.  I prefer prefix to contains as it's more performant and during an event with tons of errors we would want to this code path to be as efficient as possible. 

> Or should we map different codes in the codes package to our errors?
We need to return specific and meaningful GRPC calls so downstream will react appropriately.  Also if we hijack different grpc status codes to mean different failures due to limits we'll run out eventually.  We definitely need something more robust.

100% agree that it's not great, but I think the string error is the best way to pass custom error information back through a GRPC rpcerror.

mdisibio replied:
If nested errors are preserved across the gRPC boundary we could use `errors.Is` and const values like `errLiveTracesExceeded`

joe-elliott replied:
Sadly, it is not.  We had the same issues with the Jaeger GRPC plugin and decided to go with string matching.

mdisibio replied:
In that case it seems like string prefix is the best option.
